src/core/elf.c

```diff
@@ -40,40 +326,212 @@ elf_header_t* elf_load(uint64_t* data) {
     return elf;
 }
 
+void elf_unload(elf_header_t* elf) {
+
+    DEBUG("%s", "Unloading an elf file");
+
+    //Note: This does not currently work
+
+    elf_program_header_t* program_headers = (elf_program_header_t*)((uint64_t)elf + elf->ph_offset);
+
+    for (uint16_t i = 0; i < elf->ph_num; i++) {
+
+        elf_program_header_t* program_header = program_headers + i;
+
+        if (program_header->type == ELF_PROGRAM_TYPE_LOAD) {
+
+            DEBUG("Unmapping program header: type=%d addr=%p", program_header->type,
+                  program_header->virtual_address);
+
+            unmap(page_containing_address(program_header->virtual_address));
+        }
+    }
+
+    elf_section_header_t* sections = elf_section_header(elf);
+
+    for (uint16_t i = 0; i < elf->sh_num; i++) {
+
+        elf_section_header_t* section = &sections[i];
+
+        if (section->flags & ELF_SECTION_FLAG_ALLOC && section->size > 0) {
+
+            char* name = section->name == 0 ? NULL : elf_lookup_string(elf, section->name);
+
+            if (name == NULL) {
+
+                name = "(None)";
+            }
+
+            DEBUG("Freeing section %d (name: %d (\"%s\"), type: %d)", i, section,
+                  section->name, name, section->type);
+
+            free((void*)section->addr);
+        }
+    }
+}
+
 int is_elf(elf_header_t* elf) {
     int iself = -1;
 
     if ((elf->identity[0] == 0x7f) && !strncmp((char*)&elf->identity[1], "ELF", 3)) {
+
         iself = 0;
     }
 
+    if (iself == 0) {
+
+        DEBUG("%s", "validating elf structs");
+
+        if (elf->header_size != sizeof(elf_header_t)) {
+
+            DEBUG("%s", "Invalid elf header size");
+
+            iself = -1;
+
+        } else if (elf->ph_size != sizeof(elf_program_header_t)) {
+
+            DEBUG("%s", "Invalid program header size");
+
+            iself = -1;
+
+        } else if (elf->sh_size != sizeof(elf_section_header_t)) {
+
+            DEBUG("%s", "Invalid section header size");
+
+            iself = -1;
+        }
+    }
+
     if (iself != -1) {
+
         iself = elf->type;
     }
 
     return iself;
 }
 
-void load_segment(uint64_t* data, elf_program_header_t* program_header) {
-    uint64_t mem_size = program_header->mem_size;      // Size in memory
-    uint64_t file_size = program_header->file_size;    // Size in file
-    uint64_t addr = program_header->virtual_address;   // Offset in memory
-    uint64_t offset = program_header->offset;          // Offset in file
+#define R_AMD64_NONE 0 // None
```

Comment:
I suppose this should be located in the header file, no? Also, where do these values come from?

Reply:
They're x86_64-specific relocation operations for `REL` tables. You can find them by googling. Seems we only need 3 of them tho.